File: built-in plugins dsp/mb-synth-chiptune.cpp
/**
 * MB Chiptune
 * Category : instrument
 * Type     : analog
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : 8-bit chiptune sounds
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_SYNTH_CHIPTUNE_H
#define MB_SYNTH_CHIPTUNE_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbSynthChiptune : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-synth-chiptune";
    static constexpr const char* PLUGIN_NAME    = "MB Chiptune";
    static constexpr const char* PLUGIN_TYPE    = "analog";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float duty = 0.5f;  // range [0.1, 0.9]
    float volume = 0.7f;  // range [0, 1]
    };

    MbSynthChiptune() = default;
    ~MbSynthChiptune() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.duty = std::clamp(params.duty, 0.1f, 0.9f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Chiptune
        return input;
    }
};

#endif // MB_SYNTH_CHIPTUNE_H

File: built-in plugins dsp/mb-synth-chiptune.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbSynthChiptunePlugin: PluginDefinition = {
    id: 'mb-synth-chiptune', slug: 'mb-synth-chiptune', name: 'MB Chiptune', category: 'instrument', type: 'analog', version: '1.0.0',
    description: '8-bit chiptune sounds', author: 'Max Booster',
    oscillators: [{ type: 'square', detune: 0, gain: 0.8 }, { type: 'triangle', detune: 0, gain: 0.2 }],
    envelope: { attack: 0.001, decay: 0.1, sustain: 0.5, release: 0.1 },
    parameters: [
      { id: 'duty', name: 'Duty', type: 'float', defaultValue: 0.5, minValue: 0.1, maxValue: 0.9, automatable: true },
      { id: 'volume', name: 'Volume', type: 'float', defaultValue: 0.7, minValue: 0, maxValue: 1, automatable: true },
    ],
    defaultPreset: { duty: 0.5, volume: 0.7 },
  };

export default MbSynthChiptunePlugin;
